### ISSUE TITLE
How get event data with useEventHandler?

### ISSUE THREAD

**Issue by einarpersson** [nodegui/react-nodegui-starter#42]
I'm trying to use `useEventHandler`. I've been able to get a `QCheckBoxEvents.toggled`-event  from a `CheckBox`. However, since I have a list of checkboxes I also need to get the index (which Checkbox was checked?). The function in the eventHandlerMap seems to only get a `bool`. I tried creating a higher order function but couldn't get it to work.

What is the idea here? And what is the use of `deps` in `useEventHandler`?

```
    const checkHandler = useEventHandler({
        [QCheckBoxEvents.toggled]: () => {
            // ?
        }
    }, [])
```
```
     {things.map((v, i) => <CheckBox 
           key={i}
           text={v.text}
           checked={checkedState[i]}
           on={checkHandler}
       />)}
```

The [events-section](https://react.nodegui.org/docs/guides/handle-events) of the docs are WIP and I understand that you have loads of things to address, but I'd appreciate **any** advice on how to proceed. 

Thanks for creating this. React-nodegui seems promising!

**Comment by a7ul** [comment]
Hi
Yes, I agree there needs to be some work on the docs.
But for now just attach a console.log to see what kind of result you get via the event handler

 const checkHandler = useEventHandler({
        [QCheckBoxEvents.toggled]: console.log
    }, [])

For many useful events you would get regular javascript results of type boolean, number or string.
toggled for example gives boolean. 

But for others you would get a [External] type.  For example: KeyEvents
These [External] types mean that these are references to objects in the C++ world. 
For these you would need to import specialised handlers which are part of nodegui aswell.
For example keyevents would mean you need to do something like this: 
https://github.com/nodegui/examples/blob/193565a345649797987f3d956953aff13714a568/react-nodegui/calculator/index.tsx#L102

**Comment by a7ul** [comment]
Hope it made some sense 😅

**Comment by einarpersson** [comment]
Thanks a lot! Just realized that I put this issue in the wrong repo but it worked out anyway 

Ok, that was illuminating with `[External]`. I've encountered it but haven't known what to do with it. Now I do! :smile: 

In this particular case the problem was how to know "which checkbox was checked", which a `boolean` isn't telling. I solved it by creating a wrapping container component with a index prop (`IndexedCheckBox`). But there is probably a better way.

Thanks again!

**Comment by einarpersson** [comment]
How about this then?
```
const Test = () => {
    const [count, setCount] = useState(0)
    const handler = useEventHandler({[QPushButtonEvents.clicked]: () => {
        setCount(count + 1)
        }}, [count])
    return (
    <View>
        <Button
            on={handler}
            text='Click me'
        />
    <Text>{count}</Text>
    </View>)
}
```
It works, but gives me the following console warning every time I click the button:
`EventWidget: Couldn't unsubscribe to qt event clicked. If this is a signal you can safely ignore this warning` (alternates between _subscribe_ and _unsubscribe_)

Am I using this wrong?

**Comment by einarpersson** [comment]
(PS. I could maybe help with the docs once I get the hang of it)

**Comment by a7ul** [comment]
Hi @einarpersson 
Yes the usage is correct.

This warning is okay. 
To give you a bit of background. In Qt we have two different things to respond to an external event like a key press,etc.

1. Events - Events always are processed by the Qt's event loop.
2. Signals and slots. - Alongside events Qt also has a concept of Signals/Slots. Signals and slots are used to primarily communicate between widgets (more precisely QObjects). So the most common way of interacting between Qt Widgets is done through signals/slots. (More details here: https://doc.qt.io/qt-5/signalsandslots.html). 

Hence nodegui implements support for both events and signals.

>

> Technicals:
> An event is a message encapsulated in a class (QEvent) which is processed in an event loop and dispatched to a recipient that can either accept the message or pass it along to others to process. They are usually created in response to external system events like mouse clicks. 
> 
> Signals and Slots are a convenient way for QObjects to communicate with one another and are more similar to callback functions. In most circumstances, when a "signal" is emitted, any slot function connected to it is called directly. The exception is when signals and slots cross thread boundaries. In this case, the signal will essentially be converted into an event.


**TLDR; In JS side we have one concept of events while in Qt there are two different types - events and signal/slots. So we combine both events and signal slots via a unified interface and send it as a js event. This warning `EventWidget: Couldn't unsubscribe to qt event clicked. If this is a signal you can safely ignore this warning` is a sideeffect of that. It is just to warn a user if they intended to register a listener for an event that doesnt exist. But that would mean it would also warn for signals as well.**

**Comment by a7ul** [comment]
More details on its implementation is here: https://github.com/nodegui/nodegui/blob/master/website/docs/development/signal_and_event_handling.md

**Comment by einarpersson** [comment]
Ah, cool! Thanks for the link.

I managed to get rid of the warning by using the functional form of `setCount` and hence being able to remove the dependency on `count`, like this:
```
const Test = () => {
    const [count, setCount] = useState(0)
    const handler = useEventHandler({[QPushButtonEvents.clicked]: () => {
        setCount(c => c + 1)
        }}, [])
    return (
    <View>
        <Button
            on={handler}
            text='Click me'
        />
    <Text>{count}</Text>
    </View>)
}
```
Now it works, and I don't get any warnings.
